src/html.cr

```diff
@@ -37,9 +37,31 @@ module HTML
   # io.to_s                          # => "Crystal &amp; You"
   # ```
   def self.escape(string : String, io : IO) : Nil
-    string.each_char do |char|
-      io << SUBSTITUTIONS.fetch(char, char)
+    escape(string.to_slice, io)
+  end
+
+  # Same as `escape(String, IO)` but accepts `Bytes` instead of `String`.
+  #
+  # The slice is assumed to be valid UTF-8.
+  def self.escape(string : Bytes, io : IO) : Nil
+    last_copy_at = 0
+    string.each_with_index do |byte, index|
+      str = case byte
+            when '&'  then "&amp;"
+            when '<'  then "&lt;"
+            when '>'  then "&gt;"
+            when '"'  then "&quot;"
+            when '\'' then "&#39;"
+            else
+              nil
+            end
+      if str
+        io.write_string(string[last_copy_at, index &- last_copy_at])
+        last_copy_at = index &+ 1
+        io << str
+      end
```

Comment:
From the `else` branch we can jump directly to the next iteration, avoiding a union type for `str` and a second condition.

```suggestion
      str = case byte
            when '&'  then "&amp;"
            when '<'  then "&lt;"
            when '>'  then "&gt;"
            when '"'  then "&quot;"
            when '\'' then "&#39;"
            else
              next
            end

      io.write_string(string[last_copy_at, index &- last_copy_at])
      last_copy_at = index &+ 1
      io << str
```